source/extensions/filters/http/grpc_streaming/message_counter.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcStreaming {
+
+// The struct to store gRPC message counter state.
```

Comment:
This code already largely exists in Grpc::Frame. Please use that and/or modify if needed.

Reply:
Yes, I found it in gRPC decoder. However, it has a few issues:
- generates frame buffers as output
- does not support compressed gRPC frames
- drains the input
I suppose I could get around by mocking the inputs and outputs with fakes.  Is that what you are suggesting?

Reply:
> I suppose I could get around by mocking the inputs and outputs with fakes. Is that what you are suggesting?

No, I was just suggesting to share code if possible, like possibly the state enum, etc. Perhaps there is an opportunity to have some type of FrameInspector class or something like that? Whatever you think is best, I just don't want to duplicate the state machine.